core/src/main/java/org/mule/runtime/core/message/DefaultEventBuilder.java

```diff
@@ -389,9 +398,16 @@ private void initAfterDeserialisation(MuleContext muleContext) throws MuleExcept
         }
 
       }
-      // Can be null if service call originates from MuleClient
-      if (context.getOriginatingFlowName() != null) {
-        flowConstruct = muleContext.getRegistry().lookupFlowConstruct(context.getOriginatingFlowName());
+      if (flowName != null) {
+        flowConstruct = muleContext.getRegistry().lookupFlowConstruct(flowName);
+        // If flow construct is a Pipeline and has a EventContext instance cache, then reestablish the event context here with
+        // that instance in order to conserve non-serializable subscribers which in turn reference callbacks. Otherwise use the
+        // serialized version with no subscribers.
+        if (flowConstruct instanceof Pipeline) {
+          // TODO MULE-11349 Use unique id's for child flows
+          EventContext cachedValue = ((Pipeline) flowConstruct).getSerializedEventContextCache().remove(context.getId());
```

Comment:
make getSerializedEventContextCache an object instead of a map. Put the instanceof there, and the comment as a javadoc. This is to leave the event as clean as possible for now.